### DISCUSSION TITLE
Compiling for STM32F446ZET6 (BTT Ocotpus 1.1)

### DISCUSSION THREAD

**Discussion by ErdbeermarmeladebrotmitHonig** [grblHAL/STM32F4xx#166] [Q&A] [answered]
Hello,

currently I am trying to bring my BTT Octopus V1.1 to live. Unfortunately, after flashing the firmware it just does nothing ... . I already tried to change the clock config as discussed [here](https://github.com/grblHAL/STM32F4xx/discussions/141). Till now, I was not able to connect to the board via USB.

Is there any possibility to debug the code? I have a Nucleo Board with a ST-Link Debugger at home and even connected it to the BTT Octopus. However, I was not able to start a debug session via PlatformIO.

Can anyone help me with this?

Thanks in advance!

**Comment by ErdbeermarmeladebrotmitHonig** [comment]
Update:

I just played around with the settings and removed all the code in the function SystemClock_Config (because I thought maybe the bootloader will set everything correctly in order to use the USB). However, I had no luck on that. Then, I just added the code from [here](https://github.com/bigtreetech/BIGTREETECH-OCTOPUS-V1.0/blob/9675aed987a4ab288e3b097e09f456dc81fdb38d/Firmware/RepRapFirmware/RRFBuild-SourceCode/CoreN2G/src/STM32F4/variants/STM32F446/variant.cpp#L224-L255), but I had no luck with this as well.

I also wiped the board with the ST-Link and rewrote the bootloader. (I don't know what i was expecting from that, but I am kinda desperate right now ... :D)

In addition, I was not able to start a debug session with PlatformIO and the ST-Link debugger ...

Any help is  appreciated :)

**Reply by ErdbeermarmeladebrotmitHonig** [reply]
Oh and I just forked the repo, one can see my changes [here](https://github.com/ErdbeermarmeladebrotmitHonig/STM32F4xx/tree/master).

**Reply by terjeio** [reply]
@yoft can share his clock setup?

**Reply by ErdbeermarmeladebrotmitHonig** [reply]
Update:
I just compiled the version from @yoft and flashed it onto my board (with minor changes, I disabled everything). That code seems like it is working and an USB device shows up in my terminal. Now, I am investigating what are the differences between his code and my code.

@terjeio have you changed anything regarding USB_SERIAL_CDC or regarding the clock initialization of the STM32F446ZET6 since October? :D

**Reply by dresco** [reply]
May be a red herring, but could it be related to https://github.com/grblHAL/core/issues/472 ?

I was investigating a reported USB error on the H7 driver, but suspect it was just hardfaulting before that. Would crash consistently for me until the settings NVS had been successfully written once..

Anyway, it was fixed yesterday, so perhaps worth refreshing submodules and rebuilding?

**Reply by yoft** [reply]
Hi everyone,
I'm sorry I never got back to this, life got in the way.
I'll push my "latest" to github tonight and send a link to compare changes.

Thanks!

**Reply by ErdbeermarmeladebrotmitHonig** [reply]
@yoft that would be great!
However, your clock settings + your linker script seem like to do the trick ;)
But I am still investigating on that one :D

**Reply by ErdbeermarmeladebrotmitHonig** [reply]
Oh and @dresco thanks for letting me know!
I just pulled once again, but as i said in my last comment, I think the problem is fixed with a different linker script + setting up the clock differently ;)

**Comment by yoft** [comment]
When I did it I had trouble with the order of the clock settings, something
to do with the bootloader presetting stuff I think.

**Comment by ErdbeermarmeladebrotmitHonig** [comment]
Ok so the device now shows up in Device Manager and when calling "lsusb" from Linux. However, I am not able to interact with it via the ioSender ...

**Reply by terjeio** [reply]
> I am not able to interact with it via the ioSender ...

What exactly does this mean? Cannot connect at all? No response received error? ...

**Reply by ErdbeermarmeladebrotmitHonig** [reply]
Ok, I got it. I just forgot to call the board_init function, which is required for TMC2209.
However, now I am struggling with eeprom emul in flash and with enabling the SD-Card.

Whenever I specify the eeprom emul sections in the linker script, the board gets unresponsive.

And I wasn't able to compile the Firmware with the SDCARD flag on.

However, today was a huge step forward for me. Thanks for your help!

**Reply by ErdbeermarmeladebrotmitHonig** [reply]
Update:
Seems like I found the problem with the linker script. I am just using "STM32F446ZETX_NONVS_FLASH.ld" which was added to the repo some days ago and added the "BOOT_FLASH" memory region and the location of the VectorTable.

[Here](https://github.com/ErdbeermarmeladebrotmitHonig/STM32F4xx) is my currently working code on the BTT Octopus V1.1. However, I wasn't able to bring the SD-Card and the Flash eeprom emul to life. Hopefully I'll find time to fix this on friday.

@terjeio I saw the instructions for enabling the SD-Card in the platform.ini file. However, the files which have to be downloaded are already present. Are these instructions outdated?

**Reply by yoft** [reply, marked as answer]
Here are the changes I made to get everything working:
https://github.com/yoft/STM32F4xx/compare/master...yoft:STM32F4xx:new-octo

If you question why anything is there, it probably doesn't need to be and is left over from testing.
In the changes to main there is a section that sets clocks a second time, from memory I diagnosed that the bootloader was changing a clock, then the MCU disallowed changes straight to the desired settings, so a bit of a step around,
Good luck!

**Reply by ErdbeermarmeladebrotmitHonig** [reply]
@yoft thank you! I already took over your clock setting, it works fine.
I'll create a pull request into the original project after cleaning up my and your code if this is ok for you. In my opinion it would be great if the STM32F446ZET6 chip and the BTT Octopus 1.1/Pro boards are fully supported.

**Reply by yoft** [reply]
That would be great! I intended to do this months ago, but other thing came up, thanks for taking this on!

**Reply by yoft** [reply]
Hi mate,
Did you end up chucking in a pull request for these changes?

Thanks!

**Reply by ErdbeermarmeladebrotmitHonig** [reply]
Hi,
unfortunately not, because I have issues with the SD-Card and my TMC2209.
1. I am not able to compile the firmware with the SD_CARD flag set
2. I am not able to set other micro-stepping settings than 16 on my TMC2209. Furthermore, I was not able to set up sensorless homing.

Do you need the changes? I can share them if you want.

**Comment by ErdbeermarmeladebrotmitHonig** [comment]
@yoft I just added some changes and pushed them to my fork repo. Now, there is a common map file for the BTT Octopus Boards with a F446 chip. In my opinion it makes customization easier. I added the board file for the BTT Octopus 1.1 and updated the board file for the BTT Octopus Pro (taken from your fork).

Unfortunately, I was not able to compile with the SD Card stuff, there are always errors showing up ...

You can find my changes [here](https://github.com/ErdbeermarmeladebrotmitHonig/STM32F4xx).

**Reply by yoft** [reply]
Nice work! SD worked fine on mine, not sure what that was about. Have there been any changes to the SD code since my fork?
Did you see my changes to the flash/eeprom if/else in driver.h? From memory that made a difference, but I can't remember if I had any relevant launch parameters, I'll let you know when I get a chance.

**Reply by ErdbeermarmeladebrotmitHonig** [reply]
Thanks!

Yeah, I don't know whats the problem with that. However, I won't use the SD Card anyways so I don't care ;D
The on-board EEPROM works fine.

One thing i may change in future is the TMC2209 config. I don't know but always when I try to change the microstepping settings with the ioSender, the microsteps revert back to 16. Furthermore, I will disable stealthchop, because sometimes the motors are skipping.

Will you give that new version a try on your BTT Octopus Pro? The new board file should act like your old one because I just replaced Ports and Pins by the defines from the new map file. If yes, please let me know ;)

**Reply by ErdbeermarmeladebrotmitHonig** [reply]
@terjeio There are some known bugs in my repo (EEPROM Flash emulation doesn't work, SD-Card doesn't work, and I have troubles with setting microsteps on my TMC2209).
However, all the basic functionality (+ the Fan and Laser Plugins) work with the BTT Octopus 1.1 (untested with BTT Octopus Pro as I don't have one). Should I do a pull-request now or do you want to keep the your repo as clean as possible?

Furthermore, can you please try to compile my code with the SD Card stuff enabled (you can send me the binary afterwards, then I will test it in case you don't have a BTT Octopus)? I don't think that there is a big issue, however I wasn't able to fix it yet.

**Reply by terjeio** [reply]
> EEPROM Flash emulation doesn't work

That must be due to an incorrect loader (.ld) file since it works for me with a F466 Nucleo board.

> SD-Card doesn't work

What is the error message you get? It could be due to the wrong code beeing compiled in since both SPI and SDIO is supported?

> I have troubles with setting microsteps on my TMC2209

Is it the $-settings that change? Or the reported value from a M122 command?

>  Should I do a pull-request now or do you want to keep the your repo as clean as possible?

Are there major pin differences between the two versions? If not it might be better to combine the three map files into one and test for the board (`#ifdef BOARD_BTT_OCTOPUS_PRO ... #else ... #endif`) where there are differences.

**Reply by ErdbeermarmeladebrotmitHonig** [reply]
> That must be due to an incorrect loader (.ld) file since it works for me with a F466 Nucleo board.

Yes, I think so too, however I wasn't able to fix it. It is not a problem because the BTT Octopus has an on-board EEPROM chip.

> What is the error message you get? It could be due to the wrong code beeing compiled in since both SPI and SDIO is supported?

The first one is:
```
In file included from Src\driver.c:2204:0:
./Drivers/FATFS/App/fatfs.h:27:10: fatal error: ff_gen_drv.h: No such file or directory
```
When I change this include to `#include "FatFs/STM/ff_gen_drv.h"` I get:
```
In file included from Src\driver.c:2204:0:
./Drivers/FATFS/App/fatfs.h:28:10: fatal error: sd_diskio.h: No such file or directory
```
When I change this include to `#include "Drivers/FATFS/Target/sd_diskio.h"` I get:
```
.pio/build/btt_octopus_1_1/lib511/FatFs/ff.o: In function `sync_window.part.5':
ff.c:(.text.sync_window.part.5+0x10): undefined reference to `disk_write'
ff.c:(.text.sync_window.part.5+0x34): undefined reference to `disk_write'
.pio/build/btt_octopus_1_1/lib511/FatFs/ff.o: In function `sync_fs':
ff.c:(.text.sync_fs+0x7a): undefined reference to `disk_write'
ff.c:(.text.sync_fs+0x88): undefined reference to `disk_ioctl'
.pio/build/btt_octopus_1_1/lib511/FatFs/ff.o: In function `dir_clear':
ff.c:(.text.dir_clear+0x4c): undefined reference to `disk_write'
.pio/build/btt_octopus_1_1/lib511/FatFs/ff.o: In function `f_read':
ff.c:(.text.f_read+0x140): undefined reference to `disk_write'
.pio/build/btt_octopus_1_1/lib511/FatFs/ff.o: In function `f_write':
ff.c:(.text.f_write+0x80): undefined reference to `disk_write'
ff.c:(.text.f_write+0xc2): undefined reference to `disk_write'
.pio/build/btt_octopus_1_1/lib511/FatFs/ff.o: In function `f_sync':
ff.c:(.text.f_sync+0x2a): undefined reference to `disk_write'
.pio/build/btt_octopus_1_1/lib511/FatFs/ff.o:ff.c:(.text.f_lseek+0xde): more undefined references to `disk_write' follow
.pio/build/btt_octopus_1_1/src/driver.o: In function `sdcard_mount':
driver.c:(.text.sdcard_mount+0x6): undefined reference to `MX_FATFS_Init'
driver.c:(.text.sdcard_mount+0xe): undefined reference to `BSP_SD_Init'
collect2.exe: error: ld returned 1 exit status
*** [.pio\build\btt_octopus_1_1\firmware.elf] Error 1
```
From there I had no idea where to start  debugging.

> Is it the $-settings that change? Or the reported value from a M122 command?

Yes I think so, as I used your ioSender for configuration. And always when I hit save and then reload, the microsteps are set to 16 again.

> Are there major pin differences between the two versions? If not it might be better to combine the three map files into one and test for the board (`#ifdef BOARD_BTT_OCTOPUS_PRO ... #else ... #endif`) where there are differences.

You are right, it would make sense to only use one boards file for all BTT Octopus versions with the STM32F446ZET6 chip. I will merge the boards file for the BTT Octopus 1.1 and BTT Octopus Pro.
However, in my opinion a map file as the one I provided helps when it comes to customization. Normally, you "have to" use Motor 0 for X axis, Motor 1 for Y axis and Motor 2 for Z axis and so on. With a map file (or maybe pins file would be a better name), customization is more easy.
Currently, I am thinking about a solution where only 1 define per motor must be set (somthing like a struct). I think that would be a clean way to provide customization. What are your thoughts about that?

**Reply by terjeio** [reply]
The SD card issue is due to missing include directories/libraries in the platformio.ini file. I have not maintained it since I use the CubeIDE for compiling and do not have platformio installed on my development machines. However, I use platformio for the Web Builder and the [template file](https://github.com/grblHAL/STM32F4xx/blob/master/platformio.tpl) may contain the missing bits.

> Currently, I am thinking about a solution where only 1 define per motor must be set (somthing like a struct). I think that would be a clean way to provide customization. What are your thoughts about that?

Changing the way the map files are set up will be a lot of work, there are 50+ map files and 15 or so drivers that needs to be changed and verified. I am not willing to do this myself, but if someone else (you?) can and can come up with a good argument for why I am all ears.

**Reply by ErdbeermarmeladebrotmitHonig** [reply]
Ok thanks! I updated the platformio.ini file and now it compiles. I also removed the depricated option 'lib_extra_dirs' from all the entries in this file. Unfortunately, this took me the whole evening and thus, I had no time to test the binaries on my board yet. I still haven't done a pull-request yet, you can see the changes on [my fork](https://github.com/ErdbeermarmeladebrotmitHonig/STM32F4xx).

> Changing the way the map files are set up will be a lot of work, there are 50+ map files and 15 or so drivers that needs to be changed and verified. I am not willing to do this myself, but if someone else (you?) can and can come up with a good argument for why I am all ears.

I fully agree with you! I will think about that and will tell you as soon as I have a solution for this :D
However, I hope you got my point, that a Map-File or Pins-File or whatever it is called would make customization more easy.
